el-patch-template.el

```diff
@@ -0,0 +1,688 @@
+being patched; TYPE is a symbol `defun', `defmacro', etc."
+  (interactive (el-patch--select-template))
+  (insert (format "%S"
+                  (el-patch--resolve-template name type))))
+
+(defun el-patch-eval-template (name type)
+  "Resolve a template to an el-patch definition and evaluate it.
+
+Template should have been defined using
+`el-patch-define-template'. NAME is a symbol naming the object
+being patched; TYPE is a symbol `defun', `defmacro', etc."
+  (interactive (el-patch--select-template))
+  (eval (el-patch--resolve-template name type)))
+
+(defmacro el-patch-define-template (type-name &rest templates)
+  "Define an el-patch template.
+TYPE-NAME is a list whose first element is a type which can be
+any type from `el-patch-deftype-alist', e.g., `defun',
+`defmacro', etc, and the second element is the name of the elisp
+object to be patched or an `el-patch-*' form that resolves to
+that name. Return the new-resolved name of the object.
+
+A template in TEMPLATES can contain `...', which greedily matches
+one or more forms, and `el-patch-*' directives which are resolved
+before being matched. A template must match exactly one form in
+the definition of the elisp object, and should not match a
+subform in another template. The checks along with the actual
+matching are done when the functions `el-patch-eval-template' or
+`el-patch-insert-template' are called."
+  `(el-patch--define-template (quote ,type-name)
+                              (quote ,templates)))
+
+(defmacro el-patch-define-compiletime-template (type-name &rest templates)
+  "Define and evaluate an el-patch template.
+
+The meaning of TYPE-NAME and TEMPLATES are the same as
+`el-patch-define-template'. If called in compile-time,
+macro-expand the resolved template after defining the template.
+If called in runtime, evaluate the resolved template instead and,
+if `el-patch-warn-on-eval-template' is non-nil, print a warning."
+  (if (bound-and-true-p byte-compile-current-file)
+      (let* ((resolved-name
+              (el-patch--define-template type-name templates))
+             (resolved-template (el-patch--resolve-template
+                                 resolved-name (car type-name))))
+        `(progn
+           ;; In case el-patch-template is not loaded then simply
+           ;; set `el-patch--templates' to an empty hash
+           (setq el-patch--templates
+                 (or (bound-and-true-p el-patch--templates)
+                     (make-hash-table :test 'equal)))
+           ;; Add template definition to `el-patch--templates'
+           (puthash (quote ,(car type-name))
+                    (cons (quote ,(cadr type-name))
+                          (quote ,templates))
+                    (or (gethash (quote ,resolved-name) el-patch--templates)
+                        (puthash (quote ,resolved-name)
+                                 (make-hash-table :test #'equal)
+                                 el-patch--templates)))
```

Comment:
I figured we would want to have the template definition saved whether `el-patch-template` is loaded or not so that the user can interact with the template afterwards.

I don't know if I also need to handle the case when `el-patch` is not loaded or if that is handled by `el-patch` itself.